mava/networks/retention.py

```diff
@@ -0,0 +1,309 @@
+            xi = xi.at[:, i, :].set(xi_i)
+
+        # Repeat the decay matrix 'xi' for all agents
+        xi = jnp.repeat(xi, self.n_agents, axis=1)
+
+        return xi
+
+
+class MultiScaleRetention(nn.Module):
+    """Multi-scale retention mechanism for Sable."""
+
+    embed_dim: int
+    n_head: int
+    n_agents: int
+    memory_config: DictConfig
+    full_self_retention: bool = False
+    decay_scaling_factor: float = 1.0
+
+    def setup(self) -> None:
+        assert self.embed_dim % self.n_head == 0, "embed_dim must be divisible by n_head"
+        # Head size
+        self.head_size = self.embed_dim // self.n_head
+
+        # Decay kappa for each head
+        self.decay_kappas = 1 - jnp.exp(
+            jnp.linspace(jnp.log(1 / 32), jnp.log(1 / 512), self.n_head)
+        )
+        self.decay_kappas = self.decay_kappas * self.decay_scaling_factor
+
+        # Initialize the weights and group norm
+        self.W_G = self.param(
+            "W_G",
+            nn.initializers.normal(stddev=1 / self.embed_dim),
+            (self.embed_dim, self.head_size),
+        )
+        self.W_O = self.param(
+            "W_O",
+            nn.initializers.normal(stddev=1 / self.embed_dim),
+            (self.head_size, self.embed_dim),
+        )
+        self.group_norm = nn.GroupNorm(num_groups=self.n_head)
+
+        # Initialize the retention mechanisms
+        self.retentions = [
+            SimpleRetention(
+                self.embed_dim,
+                self.head_size,
+                self.n_agents,
+                self.full_self_retention,
+                decay_kappa,
+                self.memory_config,
+            )
+            for decay_kappa in self.decay_kappas
+        ]
+
+        # Create an instance of the positional encoding
+        self.pe = PositionalEncoding(self.memory_config, self.embed_dim)
+
+    def __call__(
+        self,
+        key: Array,
+        query: Array,
+        value: Array,
+        hstate: Array,
+        dones: Array,
+        timestep_id: Array,
+    ) -> Tuple[Array, Array]:
+        """Chunkwise (default) representation of the multi-scale retention mechanism"""
+        batch, chunk_size, _ = value.shape
+
+        # Set positional encoding
+        key, query, value = self.pe(key, query, value, timestep_id)
+
+        # Per head retention
+        ret_output = jnp.zeros((batch, chunk_size, self.head_size), dtype=value.dtype)
+        h_ns = jnp.copy(hstate)
+        for head in range(self.n_head):
+            y, h_n = self.retentions[head](key, query, value, hstate[:, head], dones)
+            ret_output = ret_output.at[
+                :, :, self.head_size * head : self.head_size * (head + 1)
+            ].set(y)
+            h_ns = h_ns.at[:, head, :, :].set(h_n)
+
+        # Gated Multi-scale retention
+        # Apply the group norm
+        ret_output = self.group_norm(ret_output.reshape(-1, self.head_size)).reshape(
+            ret_output.shape
+        )
```

Comment:
nit: no need for this comment and double reshape is confusing on 1 line
```suggestion
        ret_output = self.group_norm(ret_output.reshape(-1, self.head_size))
        ret_output = ret_output.reshape(ret_output.shape)
```